Add workflow for Cypress test on production with Chrome

MikeMcC399 opened the pull request:
This draft PR partially implements the workflow requested in https://github.com/corona-warn-app/cwa-website/issues/2892 "Add new GitHub Actions workflow carrying out the cypress test suite".

It was suggested in https://github.com/corona-warn-app/cwa-website/issues/2892#issuecomment-1141760610 to run the test regularly.

It uses https://github.com/cypress-io/github-action to run Cypress tests on the production server https://coronawarn.app once a day. The tests selected are the ones in the top level of `cypress/integration`. The problematic test `check_links.js`, which is prone to failure, is already moved down one level to `cypress/integration/hybrid`, and is not selected.

Removing the line `spec: cypress/integration/*.js` would run all tests, if that is desired.

https://github.com/cypress-io/github-action can be used as the basis for other workflows, which might also be integrated into PR submission, and it allows build and test server start parameters to be added.

I only tested this so far by manually triggering it. I did not test the `cron` part.

svengabr commented:
@MikeMcC399 thanks, we will have a look locally!

svengabr commented:
@MikeMcC399 I ran into some problems but I have fixed them in my latest commit. Now I am able to finish the workflow without errors.

https://github.com/corona-warn-app/cwa-website/commit/adb1ce970ca401a7a853f00e89d6eab6ca560839#diff-10d7404fdc00524a584d90f9b99b2e3d50d95659b03a58f21200449bbd8d43bb

![2022-06-02 15_58_40-svengabr@DESKTOP-SVEN_ _mnt_c__dev_workspaces_github_cwa-website](https://user-images.githubusercontent.com/4686874/171646546-667c6b89-0a00-4f47-917b-0842933db82f.png)

svengabr commented:
@MikeMcC399 

It's not me getting it to work, it's the Docker image that is specified to run on Ubuntu Linux `runs-on: ubuntu-latest`. All the workflow code will be executed on that said image. This Ubuntu version has only Electron installed but if we want to use Chrome instead, we have to make an effort to install chrome/chromium before the tests are executed. 

Otherwise it will fail with this error message:
![2022-06-03 08_33_20-svengabr@DESKTOP-SVEN_ _mnt_c__dev_workspaces_github_cwa-website](https://user-images.githubusercontent.com/4686874/171800126-9632fa98-4d9a-42e1-bbc9-23827d8c6b2e.png)

MikeMcC399 commented:
Good morning @svengabr!

It's curious that you get this error message, as it worked for me with `chrome` on `ubuntu-latest`. I am testing natively on GitHub in a private copy of `cwa-website`.

Here is part of the log file using chrome:

![image](https://user-images.githubusercontent.com/66998419/171809107-dae215f9-98fd-44f8-9bfd-90893b34ba89.png)



According to https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md#browsers-and-drivers `Google Chrome 102.0.5005.61` is included in `ubuntu-latest` which is currently equivalent to [ubuntu-20.04](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md). 

You mentioned in https://github.com/corona-warn-app/cwa-website/issues/2909#issuecomment-1143470698 that you were testing with https://github.com/nektos/act. I wonder if this explains the difference in test results? I haven't tried out `act` so I can't speak from personal experience. You mentioned that [ubuntu-20.04](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md) is a docker image, however I couldn't find any reference about that, although `act` definitely does say in its documentation that it is using Docker technology.

I just submitted a tweak to the script in PR https://github.com/corona-warn-app/cwa-website/pull/2945 to make the raw log output from GitHub actions more readable.

MikeMcC399 commented:
@svengabr 

> Act is running on docker as well, did you try again with my latest changes? Maybe I broke something while installing dependencies?
>I tried running on Windows 10 and on WSL2 Ubuntu and both resulted in the same problem. 

If you are having a problem with the Chrome browser I think that is most likely because your test environment does not emulate the production environment exactly. I did check again and I'm still able to run on Chrome successfully.

My test setup is a private copy of `cwa-website` where I pull the branch `cypress-workflow` from `cwa-website` and push it to my own copy. I have my private copy of the branch `cypress-workflow` set to be the default branch so I can test the workflow directly and natively. My copy of the branch `cypress-workflow` is up-to-date including https://github.com/corona-warn-app/cwa-website/commit/db153300ba162f1fcd670374d88c9cd75cd533bb.

I run the workflow through a browser to GitHub, so everything is running on GitHub. I do this from Windows 11, but that is not relevant, since the workflow runs remotely and directly in the GitHub runner.

I have attached a log file from a successful Chrome run. (The workflow is the same as the original, except I changed the browser from `electron` to `chrome`, added the environment variable `NO_COLOR` and disabled the `cron` schedule.)

[Successful Chrome run.zip](https://github.com/corona-warn-app/cwa-website/files/8831940/Successful.Chrome.run.zip)

MikeMcC399 commented:
@svengabr 

> I wonder if electron has any downside I'm comparison to chrome

I found two problems running against `electron` on GitHub which do not occur every time.

They do not cause the test to fail. It just makes the test logs look suspicious. I could not find any obvious solution for either of these issues, which is why I chose to use `chrome` instead.

## Problem 1 with Electron

I found the same error in an old open issue from 2018 https://github.com/cypress-io/cypress-docker-images/issues/52

```
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4732:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4732:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4732:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5220:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM default
[3293:0603/110439.184841:ERROR:alsa_util.cc(204)] PcmOpen: default,No such file or directory
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4732:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4732:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4732:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5220:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM default
[3293:0603/110439.193337:ERROR:alsa_util.cc(204)] PcmOpen: plug:default,No such file or directory
```

## Problem 2 with Electron

This error is listed in the open issue https://github.com/electron/electron/issues/32843

`[2984:0603/110516.880897:ERROR:node_controller.cc(567)] Trying to re-add dropped peer FBF32221E33754E2.4A8A5ACE24A3244`

---
Log attached, showing both problems:
[Electron with errors.zip](https://github.com/corona-warn-app/cwa-website/files/8831951/Electron.with.errors.zip)

MikeMcC399 commented:
@svengabr 

When you ran `act` which of the images did you choose from the ones listed on https://github.com/nektos/act#runners?

MikeMcC399 commented:
@svengabr 

I don't think that [nektos/act](https://github.com/nektos/act) should be the preferred way of testing a Cypress workflow for GitHub. It is better to test directly in GitHub on a copy of the repository.

Cypress is for testing against web browsers. The default `act` Medium Docker Image `ghcr.io/catthehacker/ubuntu:act-20.04` however does not contain any web browsers. The `act` Large Docker Image `ghcr.io/catthehacker/ubuntu:full-20.04` is built from the repository https://github.com/catthehacker/virtual-environments-fork, which is forked from the GitHub repository https://github.com/actions/virtual-environments/. The `catthehacker` fork is however left at a commit level of September 2021, which is 9 months behind GitHub. So, neither `act` image corresponds to the current virtual environment [Ubuntu 20.04.4 LTS](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md) used by GitHub, which **does** contain up-to-date [Browsers and Drivers](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md#browsers-and-drivers):

- Google Chrome 102.0.5005.61
- ChromeDriver 102.0.5005.61
- Chromium 102.0.5005.0
- Selenium server 4.2.0
- Mozilla Firefox 100.0.2
- Geckodriver 0.31.0

[nektos/act](https://github.com/nektos/act) also introduces additional dependencies of Docker, WSL2 / Hyper-V or Ubuntu plus other package managers.

[nektos/act: Default runners are intentionally incomplete](https://github.com/nektos/act#default-runners-are-intentionally-incomplete) says: "These default images do **not** contain **all** the tools that GitHub Actions offers by default in their runners. Many things can work improperly or not at all while running those image. Additionally, some software might still not work even if installed properly, since GitHub Actions are running in fully virtualized machines while `act` is using Docker containers (e.g. Docker does not support running `systemd`)."

I have only a small amount of experience with Docker and I found it very difficult to get everything running, with many individual steps, so for me it was a high overhead to set up. I was not able to pull the full Docker image (around 20GB) as the download either hung up or failed with `unexpected EOF` after having downloaded several Gigabytes.

So, although it was very interesting to look at [nektos/act](https://github.com/nektos/act) I suggest to drop its use for Cypress testing. If a Cypress test run fails on `act` it does not mean that it will fail on GitHub (e.g. Chrome works on GitHub and it is not necessary to install additional Ubuntu dependencies). If a Cypress test run is successful on `act` there is no guarantee that it will work on GitHub because the contents and version levels are not aligned between `act` and GitHub.

svengabr commented:
> I don't think that [nektos/act](https://github.com/nektos/act) should be the preferred way of testing a Cypress workflow

@MikeMcC399 yes I agree with that. I will as well change from act to my own copy. Also, my colleague @brifemu is not available this week so I will take care of cypress in the meantime.
